File: packages/fliwright-core/tests/AppInstance.test.ts
import { describe, expect, it } from 'vitest';
import { FliwrightDriver } from '../src/Driver.js';
import { createProtocolMock } from './helpers/mockVMService.js';

describe('AppInstance', () => {
  it('reads app info through the bridge app extension', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.info', (params) => {
      expect(params.isolateId).toBe(protocol.isolateId);
      return {
        id: 'exio',
        name: 'Exio',
        environment: 'dev',
        capabilities: ['auth', 'exio.account'],
      };
    });
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    const info = await driver.app.info();

    expect(info).toEqual({
      id: 'exio',
      name: 'Exio',
      environment: 'dev',
      capabilities: ['auth', 'exio.account'],
    });
  });

  it('reads app snapshots without assuming business-specific fields', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.snapshot', () => ({
      id: 'shop',
      capabilities: ['auth'],
      snapshot: {
        route: '/checkout',
        auth: { isAuthenticated: true, userId: 'u_1' },
        cart: { items: 2 },
      },
    }));
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    const snapshot = await driver.app.getSnapshot<{
      route: string;
      auth: { isAuthenticated: boolean; userId: string };
      cart: { items: number };
    }>();

    expect(snapshot.snapshot.route).toBe('/checkout');
    expect(snapshot.snapshot.auth.userId).toBe('u_1');
    expect(snapshot.snapshot.cart.items).toBe(2);
  });

  it('lists and invokes capabilities with JSON input', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.capabilities', () => ({
      capabilities: [
        { name: 'auth', description: 'Authentication state', methods: ['getStatus'] },
      ],
    }));
    protocol.mockExtension('ext.fliwright.app.invoke', (params) => {
      expect(params.capability).toBe('auth');
      expect(params.method).toBe('getStatus');
      expect(JSON.parse(params.input)).toEqual({ refresh: true });
      return {
        success: true,
        result: { isAuthenticated: true, userId: 'u_1' },
      };
    });
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    await expect(driver.app.listCapabilities()).resolves.toEqual([
      { name: 'auth', description: 'Authentication state', methods: ['getStatus'] },
    ]);
    await expect(
      driver.app.invoke('auth', 'getStatus', { refresh: true }),
    ).resolves.toEqual({ isAuthenticated: true, userId: 'u_1' });
  });

  it('returns undefined for missing capabilities', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.info', () => ({
      id: 'plain-app',
      capabilities: [],
    }));
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    await expect(driver.app.getCapability('auth')).resolves.toBeUndefined();
  });

  it('maps typed capability methods to bridge invocation', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.info', () => ({
      id: 'exio',
      capabilities: ['auth'],
    }));
    protocol.mockExtension('ext.fliwright.app.invoke', (params) => {
      expect(params.capability).toBe('auth');
      expect(params.method).toBe('getStatus');
      return {
        success: true,
        result: { isAuthenticated: true, userId: 'u_1' },
      };
    });
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    const auth = await driver.app.getCapability<{
      getStatus(): Promise<{ isAuthenticated: boolean; userId?: string }>;
    }>('auth');

    await expect(auth?.getStatus()).resolves.toEqual({
      isAuthenticated: true,
      userId: 'u_1',
    });
  });

  it('throws when a capability invocation returns an extension error payload', async () => {
    const protocol = createProtocolMock();
    protocol.mockExtension('ext.fliwright.app.invoke', () => ({
      success: false,
      error: 'Capability "auth" is not registered',
    }));
    const driver = new FliwrightDriver();
    await driver.attachMockConnector(protocol.ws);

    await expect(driver.app.invoke('auth', 'getStatus')).rejects.toThrow(
      'Capability "auth" is not registered',
    );
  });
});

File: packages/fliwright-bridge/test/app_instance_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:fliwright_bridge/fliwright_bridge.dart';

void main() {
  setUp(() async {
    await FliwrightBridge.reset();
  });

  tearDown(() async {
    await FliwrightBridge.reset();
  });

  test(
    'init registers app instance extensions and advertises capability',
    () async {
      await FliwrightBridge.initForDioMock();

      expect(
        FliwrightBridge.registry.registeredMethods,
        containsAll([
          'ext.fliwright.app.info',
          'ext.fliwright.app.snapshot',
          'ext.fliwright.app.capabilities',
          'ext.fliwright.app.invoke',
        ]),
      );

      final handshake = await FliwrightBridge.registry.invoke(
        'ext.fliwright.handshake',
        {'protocolVersion': '1'},
      );
      final capabilities =
          handshake['bridgeCapabilities'] as Map<dynamic, dynamic>;
      expect(capabilities['appInstance'], isTrue);
      expect(capabilities['appCapabilities'], isTrue);
    },
  );

  test('app info and snapshot are generic and app-defined', () async {
    FliwrightAppInstance.configure(
      id: 'exio',
      name: 'Exio',
      environment: 'dev',
      snapshot: () => {
        'route': '/home',
        'auth': {'isAuthenticated': true, 'userId': 'u_1'},
        'featureFlags': {'newCheckout': true},
      },
    );
    FliwrightAppInstance.registerCapability(
      FliwrightAppCapability(
        name: 'auth',
        description: 'Authentication state',
        methods: {
          'getStatus': (_) => {'isAuthenticated': true},
        },
      ),
    );
    await FliwrightBridge.initForDioMock();

    final info = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.info',
      {},
    );
    expect(info['id'], 'exio');
    expect(info['name'], 'Exio');
    expect(info['environment'], 'dev');
    expect(info['capabilities'], contains('auth'));

    final snapshot = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.snapshot',
      {},
    );
    expect(snapshot['id'], 'exio');
    expect(snapshot['snapshot'], {
      'route': '/home',
      'auth': {'isAuthenticated': true, 'userId': 'u_1'},
      'featureFlags': {'newCheckout': true},
    });
  });

  test('capability descriptors and invocation use JSON input', () async {
    FliwrightAppInstance.registerCapability(
      FliwrightAppCapability(
        name: 'auth',
        description: 'Authentication state',
        methods: {
          'getStatus': (input) => {'isAuthenticated': true, 'input': input},
        },
      ),
    );
    await FliwrightBridge.initForDioMock();

    final capabilities = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.capabilities',
      {},
    );
    expect(capabilities['capabilities'], [
      {
        'name': 'auth',
        'description': 'Authentication state',
        'methods': ['getStatus'],
      },
    ]);

    final result = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.invoke',
      {
        'capability': 'auth',
        'method': 'getStatus',
        'input': jsonEncode({'refresh': true}),
      },
    );
    expect(result['success'], isTrue);
    expect(result['result'], {
      'isAuthenticated': true,
      'input': {'refresh': true},
    });
  });

  test(
    'invoke reports missing capability without throwing through registry',
    () async {
      await FliwrightBridge.initForDioMock();

      final result = await FliwrightBridge.registry.invoke(
        'ext.fliwright.app.invoke',
        {'capability': 'auth', 'method': 'getStatus'},
      );

      expect(result['success'], isFalse);
      expect(result['error'], contains('Capability "auth" is not registered'));
    },
  );

  test('auth capability standardizes seeded login handlers', () async {
    final calls = <Object?>[];
    FliwrightAppInstance.registerCapability(
      FliwrightAuthCapability(
        seedLoggedIn: (input) {
          calls.add(input);
          return {
            'isAuthenticated': true,
            'userId': (input as Map)['userId'],
          };
        },
        clearSession: (_) => {'isAuthenticated': false},
      ),
    );
    await FliwrightBridge.initForDioMock();

    final capabilities = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.capabilities',
      {},
    );
    expect(capabilities['capabilities'], [
      {
        'name': 'auth',
        'description': 'Authentication test capability',
        'methods': ['seedLoggedIn', 'clearSession'],
      },
    ]);

    final seeded = await FliwrightBridge.registry.invoke(
      'ext.fliwright.app.invoke',
      {
        'capability': 'auth',
        'method': 'seedLoggedIn',
        'input': jsonEncode({'userId': 'u_1'}),
      },
    );

    expect(seeded['success'], isTrue);
    expect(seeded['result'], {'isAuthenticated': true, 'userId': 'u_1'});
    expect(calls, [
      {'userId': 'u_1'},
    ]);
  });

  test('auth capability requires at least one handler', () {
    expect(() => FliwrightAuthCapability(), throwsArgumentError);
  });
}
